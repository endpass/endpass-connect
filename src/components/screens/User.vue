<template>
  <screen>
    <v-frame :loading="!inited">
      <account-form
        :loading="loading"
        :closable="isDialog"
        :accounts="accountsOptions"
        :networks="networksOptions"
        :form-data="formData"
        :error="error"
        :message="message"
        @donate-request="handleDonateRequest"
        @donate-success="handleDonateSuccess"
        @donate-error="handleDonateError"
        @submit="handleAccountFormSubmit"
        @cancel="handleCancel"
        @logout="handleLogout"
      />
    </v-frame>
  </screen>
</template>

<script>
import { mapActions, mapMutations, mapState, mapGetters } from 'vuex';
import { DEFAULT_NETWORKS } from '@/constants';
import Screen from '../Screen.vue';
import VFrame from '../VFrame.vue';
import AccountForm from '../forms/Account.vue';

export default {
  name: 'User',

  data: () => ({
    formData: {
      activeAccount: null,
      activeNet: null,
    },
    error: null,
    message: null,
  }),

  computed: {
    ...mapState({
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      settings: state => state.accounts.settings,
    }),
    ...mapGetters(['availableAccounts', 'isDialog']),

    networksOptions() {
      return Object.values(DEFAULT_NETWORKS).map(({ id, name }) => ({
        value: id,
        label: name,
      }));
    },

    accountsOptions() {
      return this.availableAccounts.map(({ address }) => ({
        value: address,
        label: address,
      }));
    },
  },

  watch: {
    formData: {
      handler() {
        if (this.message) {
          this.message = null;
        }

        if (this.error) {
          this.error = null;
        }
      },
      deep: true,
    },
  },

  methods: {
    ...mapMutations(['changeLoadingStatus']),
    ...mapActions(['logout', 'closeAccount', 'getAccounts', 'updateSettings']),

    async handleAccountFormSubmit() {
      const { activeAccount, activeNet } = this.formData;

      try {
        this.error = null;

        await this.updateSettings({
          lastActiveAccount: activeAccount,
          net: activeNet,
        });
      } catch (err) {
        console.error(err);

        this.error = err.message;
      }
    },

    async handleLogout() {
      this.logout();
    },

    handleCancel() {
      this.closeAccount();
    },

    handleDonateRequest() {
      this.changeLoadingStatus(true);
    },

    handleDonateSuccess() {
      this.message =
        'Donation request successfully sent. In the most cases it can takes a few time before you will receive funds.';
      this.changeLoadingStatus(false);
    },

    handleDonateError(e) {
      this.error = e;
      this.changeLoadingStatus(false);
    },

    handleWindowClose() {
      this.closeAccount();
    },
  },

  async created() {
    const { settings } = this;

    this.formData.activeAccount = settings.lastActiveAccount;
    this.formData.activeNet = settings.net;

    if (this.isDialog) {
      window.addEventListener('beforeunload', this.handleWindowClose);
    }
  },

  components: {
    Screen,
    VFrame,
    AccountForm,
  },
};
</script>
