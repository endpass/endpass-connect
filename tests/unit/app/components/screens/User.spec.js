import Vuex from 'vuex';
import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import User from '@/components/screens/User.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('User', () => {
  let store;
  let storeData;
  let wrapper;
  let accountsModule;
  let coreModule;

  beforeEach(() => {
    coreModule = {
      state: {
        inited: true,
        loading: false,
      },
      getters: {
        isDialog: jest.fn(() => true),
      },
    };
    accountsModule = {
      state: {
        linkSent: false,
        accounts: null,
        settings: {
          lastActiveAccount: '0x0',
          net: 1,
        },
      },
      actions: {
        logout: jest.fn(),
        closeAccount: jest.fn(),
        updateSettings: jest.fn(),
        getAccounts: jest.fn(),
      },
      getters: {
        availableAccounts: jest.fn(() => [
          {
            address: '0x0',
            type: 'StandardAccount',
          },
        ]),
      },
    };
    storeData = {
      modules: {
        accounts: accountsModule,
        core: coreModule,
      },
    };
    store = new Vuex.Store(storeData);
    wrapper = shallowMount(User, {
      localVue,
      store,
    });
  });

  describe('render', () => {
    it('should correctly render User component', () => {
      expect(wrapper.name()).toBe('User');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should set form data from settings on create', () => {
      expect(wrapper.vm.formData).toEqual({
        activeAccount: '0x0',
        activeNet: 1,
      });
    });

    it('should update settings of form submit', () => {
      const { activeAccount, activeNet } = wrapper.vm.formData;

      wrapper.find('account-form-stub').vm.$emit('submit');

      expect(accountsModule.actions.updateSettings).toBeCalledWith(
        expect.any(Object),
        {
          lastActiveAccount: activeAccount,
          net: activeNet,
        },
        undefined,
      );
    });

    it('should logout if logout button was pressed in form', () => {
      wrapper.find('account-form-stub').vm.$emit('logout');

      expect(accountsModule.actions.logout).toBeCalled();
    });

    it('should close account if cancel button was pressed in form', () => {
      wrapper.find('account-form-stub').vm.$emit('cancel');

      expect(accountsModule.actions.closeAccount).toBeCalled();
    });
  });

  // describe('watchers', () => {
  //   it('should do something', () => {
  //     expect('1 + 1').toBe(2)
  //   })
  // })
});
