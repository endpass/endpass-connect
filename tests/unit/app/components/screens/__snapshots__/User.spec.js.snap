// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User render should correctly render User component 1`] = `
<screen-stub>
  <v-frame-stub>
    <account-form-stub closable="true" accounts="[object Object]" networks="[object Object],[object Object],[object Object],[object Object]" formdata="[object Object]"></account-form-stub>
  </v-frame-stub>
</screen-stub>
`;
